bugsnag-android-core/src/main/java/com/bugsnag/android/DeviceIdStore.kt

```diff
@@ -0,0 +1,179 @@
+package com.bugsnag.android
+
+import android.content.Context
+import android.util.JsonReader
+import java.io.File
+import java.io.IOException
+import java.lang.Thread
+import java.nio.channels.FileChannel
+import java.nio.channels.FileLock
+import java.nio.channels.OverlappingFileLockException
+import java.util.UUID
+
+/**
+ * This class is responsible for persisting and retrieving the device ID which uniquely
+ * identifies this device.
+ */
+internal class DeviceIdStore @JvmOverloads constructor(
+    private val context: Context,
+    private val file: File = File(context.filesDir, "device-id"),
+    private val logger: Logger
+) {
+
+    private val synchronizedStreamableStore: SynchronizedStreamableStore<DeviceId>
+
+    init {
+        try {
+            if (!file.exists()) {
+                file.createNewFile()
+            }
+        } catch (exc: IOException) {
+            logger.w("Failed to created device ID file", exc)
+        }
+        this.synchronizedStreamableStore = SynchronizedStreamableStore(file)
+    }
+
+    /**
+     * Loads the device ID from its file system location. Device IDs are UUIDs which are
+     * persisted on a per-install basis. This method is thread-safe and multi-process safe.
+     *
+     * If no device ID exists then the legacy value stored in [SharedPreferences] will
+     * be used. If no value is present then a random UUID will be generated and persisted.
+     */
+    fun loadDeviceId(): String? {
+        val sharedPrefMigrator = SharedPrefMigrator(context)
+        val legacyDeviceId = sharedPrefMigrator.loadDeviceId()
+
+        return loadDeviceId {
+            when (legacyDeviceId) {
+                null -> UUID.randomUUID()
+                else -> UUID.fromString(legacyDeviceId)
+            }
+        }
+    }
+
+    internal fun loadDeviceId(uuidProvider: () -> UUID): String? {
+        return try {
+            // optimistically read device ID without a lock - the majority of the time
+            // the device ID will already be present so no synchronization is required.
+            val deviceId = loadDeviceIdInternal()
+
+            if (deviceId?.id != null) {
+                deviceId.id
+            } else {
+                return persistNewDeviceUuid(uuidProvider)
+            }
+        } catch (exc: Throwable) {
```

Comment:
Is there good cause to catch `Throwable` here? I would be keen to avoid this and catch more specific errors as we don't catch throwables elsewhere in the codebase. (Also the catch in `persistNewDeviceUuid` is dong the same an so seem redundant.)